### PR TITLE
MULTIARCH-3882: Support PER solution for powervs platform via transit gateway

### PR DESCRIPTION
**Pull request by dharaneeshvrd**
More about PER: https://cloud.ibm.com/docs/power-iaas?topic=power-iaas-per

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by csrwng** [comment]
Can you please separate the vendor file changes into a separate commit?

File: cmd/cluster/core/create.go [hunk 1]
```diff
@@ -91,16 +91,19 @@ type CreateOptions struct {
...
+	Region                 string
+	Zone                   string
+	CloudInstanceID        string
+	CloudConnection        string
+	VPCRegion              string
+	VPC                    string
+	VPCSubnet              string
+	Debug                  bool
+	RecreateSecrets        bool
+	UsePER                 bool
```
Comment (mkumatag) [inline cmd/cluster/core/create.go:104]: These fields are increasing, lets start adding comments for these fields to have better understanding

File: cmd/cluster/powervs/create.go [hunk 2]
```diff
@@ -50,6 +51,9 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().Int32Var(&opts.PowerVSPlatform.Memory, "memory", opts.PowerVSPlatform.Memory, "Amount of memory allocated (in GB). Default is 32")
 	cmd.Flags().BoolVar(&opts.PowerVSPlatform.Debug, "debug", opts.PowerVSPlatform.Debug, "Enabling this will print PowerVS API Request & Response logs")
 	cmd.Flags().BoolVar(&opts.PowerVSPlatform.RecreateSecrets, "recreate-secrets", opts.PowerVSPlatform.RecreateSecrets, "Enabling this flag will recreate creds mentioned https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1alpha1.PowerVSPlatformSpec here. This is required when rerunning 'hypershift create cluster powervs' or 'hypershift create infra powervs' commands, since API key once created cannot be retrieved again. Please make sure that cluster name used is unique across different management clusters before using this flag")
+	cmd.Flags().BoolVar(&opts.PowerVSPlatform.UsePER, "use-per", opts.PowerVSPlatform.UsePER, "Enabling this flag will utilize PER solution via transit gateway instead of cloud connection to create a connection between PowerVS and VPC")
```
Comment (mkumatag) [inline cmd/cluster/powervs/create.go:54]: can we change the name to `power-edge-router`

File: cmd/cluster/powervs/destroy.go [hunk 3]
```diff
@@ -39,12 +40,16 @@ func NewDestroyCommand(opts *core.DestroyOptions) *cobra.Command {
 	cmd.Flags().StringVar(&opts.PowerVSPlatform.CloudInstanceID, "cloud-instance-id", "", "IBM Cloud PowerVS Service Instance ID. Use this flag to reuse an existing PowerVS Service Instance resource for cluster's infra")
 	cmd.Flags().StringVar(&opts.PowerVSPlatform.CloudConnection, "cloud-connection", "", "Cloud Connection in given zone. Use this flag to reuse an existing Cloud Connection resource for cluster's infra")
 	cmd.Flags().BoolVar(&opts.PowerVSPlatform.Debug, "debug", opts.PowerVSPlatform.Debug, "Enabling this will print PowerVS API Request & Response logs")
+	cmd.Flags().BoolVar(&opts.PowerVSPlatform.UsePER, "use-per", opts.PowerVSPlatform.UsePER, "Enabling this flag will utilize PER solution via transit gateway instead of cloud connection to create a connection between PowerVS and VPC")
```
Comment (mkumatag) [inline cmd/cluster/powervs/destroy.go:43]: same comment as above

File: cmd/infra/powervs/create.go [hunk 4]
```diff
@@ -218,12 +228,16 @@ func NewCreateCommand() *cobra.Command {
 	cmd.Flags().StringVar(&opts.OutputFile, "output-file", opts.OutputFile, "Path to file that will contain output information from infra resources (optional)")
 	cmd.Flags().BoolVar(&opts.Debug, "debug", opts.Debug, "Enabling this will print PowerVS API Request & Response logs")
 	cmd.Flags().BoolVar(&opts.RecreateSecrets, "recreate-secrets", opts.RecreateSecrets, "Enabling this flag will recreate creds mentioned https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1alpha1.PowerVSPlatformSpec here. This is required when rerunning 'hypershift create cluster powervs' or 'hypershift create infra powervs' commands, since API key once created cannot be retrieved again. Please make sure that cluster name used is unique across different management clusters before using this flag")
+	cmd.Flags().BoolVar(&opts.UsePER, "use-per", opts.UsePER, "Enabling this flag will utilize PER solution via transit gateway instead of cloud connection to create a connection between PowerVS and VPC")
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:231]: same as above

File: cmd/infra/powervs/create.go [hunk 5]
```diff
@@ -1438,3 +1470,128 @@ func attachTag(gtag *globaltaggingv1.GlobalTaggingV1, infraID string, resourceId
...
+			return fmt.Errorf("error creating transit gateway: %w", err)
+		}
+	}
+
+	if transitGateway != nil {
+		infra.TransitGatewayID = *transitGateway.ID
+		infra.Stats.TransitGatewayState.Status = *transitGateway.Status
+	} else {
+		return fmt.Errorf("unable to setup transit gateway")
+	}
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1505]:
```suggestion
	if transitGateway == nil {
		return fmt.Errorf("unable to setup transit gateway")
	}
	infra.TransitGatewayID = *transitGateway.ID
	infra.Stats.TransitGatewayState.Status = *transitGateway.Status
```

File: cmd/infra/powervs/create.go [hunk 6]
```diff
@@ -1438,3 +1470,128 @@ func attachTag(gtag *globaltaggingv1.GlobalTaggingV1, infraID string, resourceId
...
+// createTransitGateway creates transit gateway and connections
+func (infra *Infra) createTransitGateway(ctx context.Context, options *CreateInfraOptions, tgapisv1 *transitgatewayapisv1.TransitGatewayApisV1) (*transitgatewayapisv1.TransitGateway, error) {
+
+	transitGatewayName := fmt.Sprintf("%s-%s", options.InfraID, transitGatewayNameSuffix)
+
+	tg, err := validateTransitGatewayByName(ctx, tgapisv1, transitGatewayName, true)
+	if err != nil && err.Error() != transitGatewayNotFound(transitGatewayName).Error() {
+		return nil, fmt.Errorf("error validating existing transit gateway: %w", err)
+	}
+	if tg != nil && *tg.Status == "available" {
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1524]: Is there a scenario where tg is not nil and status is not in `available` state then what happens? Ideally even a create call also should fail as tg already exists, then how do we handle that scenario?

File: cmd/infra/powervs/create.go [hunk 7]
```diff
@@ -1438,3 +1470,128 @@ func attachTag(gtag *globaltaggingv1.GlobalTaggingV1, infraID string, resourceId
...
+		log(options.InfraID).Info("Using existing transit gateway ...")
+		return tg, nil
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1526]: if available the does it mean that we also have connections as well? what if we just have the transit gateway but not the connections for vpc and powervs?

File: cmd/infra/powervs/destroy.go [hunk 8]
```diff
@@ -99,6 +103,9 @@ func NewDestroyCommand() *cobra.Command {
 	cmd.Flags().StringVar(&opts.CloudConnection, "cloud-connection", opts.CloudConnection, "IBM Cloud PowerVS Cloud Connection. Use this flag to reuse an existing Cloud Connection resource for cluster's infra")
 	cmd.Flags().StringVar(&opts.CloudInstanceID, "cloud-instance-id", opts.CloudInstanceID, "IBM PowerVS Cloud Instance ID. Use this flag to reuse an existing PowerVS Cloud Instance resource for cluster's infra")
 	cmd.Flags().BoolVar(&opts.Debug, "debug", opts.Debug, "Enabling this will result in debug logs will be printed")
+	cmd.Flags().BoolVar(&opts.UsePER, "use-per", opts.UsePER, "Enabling this flag will utilize PER solution via transit gateway instead of cloud connection to create a connection between PowerVS and VPC")
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:106]: same as above

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/infra/powervs/create.go (hunk 6 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go:1524]: Yes, when `validateTransitGatewayByName` returns transitGatewayNotFound error, tg would be nil, so before accessing `*tg.status`, need to validate it's not nil. If its nil, will proceed to create a new transit gatway. So we need a nil check here. We will not attempt the create if its already exists.

File: cmd/infra/powervs/create.go (hunk 7 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go:1526]: I am validating the vpc and powervs connections are attached to this tg and their status is ok in `validateTransitGatewayByName` func. Then only I would return the tg obj and nil. If tg is there and connections are not proper then I return `transitGatewayConnectionError`.

**Review by dharaneeshvrd — COMMENTED** [review]
@mkumatag 
I have addressed your other comments too, please take a look!

File: cmd/infra/powervs/create.go (hunk 6 above)
Comment (mkumatag) [reply cmd/infra/powervs/create.go:1524]: what if tg is not null(transit gateway exist) and state is not `available`?

File: cmd/infra/powervs/destroy.go [hunk 9]
```diff
@@ -99,6 +103,9 @@ func NewDestroyCommand() *cobra.Command {
 	cmd.Flags().StringVar(&opts.CloudConnection, "cloud-connection", opts.CloudConnection, "IBM Cloud PowerVS Cloud Connection. Use this flag to reuse an existing Cloud Connection resource for cluster's infra")
 	cmd.Flags().StringVar(&opts.CloudInstanceID, "cloud-instance-id", opts.CloudInstanceID, "IBM PowerVS Cloud Instance ID. Use this flag to reuse an existing PowerVS Cloud Instance resource for cluster's infra")
 	cmd.Flags().BoolVar(&opts.Debug, "debug", opts.Debug, "Enabling this will result in debug logs will be printed")
+	cmd.Flags().BoolVar(&opts.UsePER, "use-power-edge-router", opts.UsePER, "Enabling this flag will denotes that Power Edge Router solution is used while creating the cluster")
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:106]:
```suggestion
	cmd.Flags().BoolVar(&opts.PER, "power-edge-router", opts.PER, "Enabling this flag will denotes that Power Edge Router solution is used while creating the cluster")
```

File: cmd/infra/powervs/validate.go [hunk 10]
```diff
@@ -204,3 +209,62 @@ func validateCloudConnectionInPowerVSZone(name string, client *instance.IBMPIClo
...
+
+	if err := pagingHelper(f); err != nil {
+		return nil, err
+	}
+
+	if transitGateway == nil {
+		return nil, transitGatewayNotFound(name)
+	}
+
+	if validateStatus {
```
Comment (mkumatag) [inline cmd/infra/powervs/validate.go:246]: you can have early return here in case if `validateStatus` is set as false

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Review by mkumatag — APPROVED** [review]
/lgtm

File: cmd/infra/powervs/destroy.go [hunk 11]
```diff
@@ -99,6 +103,9 @@ func NewDestroyCommand() *cobra.Command {
 	cmd.Flags().StringVar(&opts.CloudConnection, "cloud-connection", opts.CloudConnection, "IBM Cloud PowerVS Cloud Connection. Use this flag to reuse an existing Cloud Connection resource for cluster's infra")
 	cmd.Flags().StringVar(&opts.CloudInstanceID, "cloud-instance-id", opts.CloudInstanceID, "IBM PowerVS Cloud Instance ID. Use this flag to reuse an existing PowerVS Cloud Instance resource for cluster's infra")
 	cmd.Flags().BoolVar(&opts.Debug, "debug", opts.Debug, "Enabling this will result in debug logs will be printed")
+	cmd.Flags().BoolVar(&opts.PER, "power-edge-router", opts.PER, "Enabling this flag will denotes that Power Edge Router solution is used while creating the cluster")
```
Comment (csrwng) [inline cmd/infra/powervs/destroy.go:106]: The description doesn't seem to make sense for the destroy command. Something like, "Enabling this flag ensures that the Power Edge router that was used to create the cluster is cleaned up."

File: cmd/infra/powervs/destroy.go [hunk 12]
```diff
@@ -99,6 +103,9 @@ func NewDestroyCommand() *cobra.Command {
...
+	cmd.Flags().StringVar(&opts.TransitGatewayLocation, "transit-gateway-location", opts.TransitGatewayLocation, "IBM Cloud Transit Gateway location")
+	cmd.Flags().StringVar(&opts.TransitGateway, "transit-gateway", opts.TransitGateway, "IBM Cloud Transit Gateway. Use this flag to reuse an existing Transit Gateway resource for cluster's infra")
```
Comment (csrwng) [inline cmd/infra/powervs/destroy.go:108]: Same as above, the wording should be different for the destroy command.

File: cmd/infra/powervs/destroy.go [hunk 13]
```diff
@@ -908,3 +927,73 @@ func deleteCOS(ctx context.Context, options *DestroyInfraOptions, resourceGroupI
...
+func destroyTransitGateway(ctx context.Context, options *DestroyInfraOptions) error {
+	tgapisv1, err := transitgatewayapisv1.NewTransitGatewayApisV1(&transitgatewayapisv1.TransitGatewayApisV1Options{
+		Authenticator: getIAMAuth(),
+		Version:       utilpointer.String(currentDate),
+	})
+	if err != nil {
+		return err
+	}
+
+	if options.TransitGateway == "" {
```
Comment (csrwng) [inline cmd/infra/powervs/destroy.go:941]: nit: return early instead of enclosing a large block in the if

**Comment by csrwng** [comment]
Some comments, otherwise lgtm
/approve

**Pull request merged**